Futura/Futura/Future.swift

```diff
@@ -0,0 +1,327 @@
+                        try future.become(.resulted(with: .success(transformation(val))))
+                    } catch {
+                        future.become(.resulted(with: .error(error)))
+                    }
+                case let .error(err):
+                    future.become(.resulted(with: .error(err)))
+                }
+            case .canceled:
+                future.become(.canceled)
+            case .waiting: break
+            }
+        }
+        return future
+    }
+    
+    /// Map container to other type or do some value changes flattening future inside. Returns new Future instance.
+    /// Transformation may throw to propagate error instad of value.
+    func flatMap<T>(_ transformation: @escaping (Value) throws -> (Future<T>)) -> Future<T> {
+        let future = Future<T>(executionContext: .inheritWorker)
+        #if FUTURA_DEBUG
+        os_log("Flat mapping on %{public}@ => %{public}@", log: logger, type: .debug, debugDescriptionSynchronized, future.debugDescriptionSynchronized)
+        #endif
+        observe { state in
+            switch state {
+            case let .resulted(result):
+                switch result {
+                case let .success(val):
+                    do {
+                        try transformation(val).handle {
+                            future.become(.resulted(with: $0))
+                        }
+                    } catch {
+                        future.become(.resulted(with: .error(error)))
+                    }
+                case let .error(err):
+                    future.become(.resulted(with: .error(err)))
+                }
+            case .canceled:
+                future.become(.canceled)
+            case .waiting: break
+            }
+        }
+        return future
+    }
+    
+    /// Returns new Future instance with given execution context. Futures inherit execution context by default.
+    func `switch`(to worker: Worker) -> Future<Value> {
+        let future = Future(executionContext: .async(using: worker))
+        #if FUTURA_DEBUG
+        os_log("Switching to %{public}@ on %{public}@", log: logger, type: .debug, String(describing: worker) as! CVarArg, debugDescriptionSynchronized)
+        #endif
+        observe { future.become($0) }
+        return future
+    }
+    
+    /// Cancels future without triggering any handlers (except always and canceled). Cancellation is propagated.
+    func cancel() {
+        become(.canceled)
+    }
+}
+
+internal extension Future {
+    
+    enum State {
+        case waiting
+        case resulted(with: Result<Value>)
+        case canceled
+    }
+    
+    func become(_ state: State) {
+        lock.synchronized {
+            switch self.state {
+            case .resulted, .canceled:
+                #if FUTURA_DEBUG
+                os_log("Ignoring completion with %{public}@ on %{public}@", log: logger, type: .debug, String(describing: state) as! CVarArg, debugDescription)
```

Comment:
This 
```
#if FUTURA DEBUG
os_log(...)
#endif``` gets repeated from time to time, maybe consider some small extraction of the logging mechanism? :P I know it makes not much sense but it obscures the code a little bit

Reply:
I know. This obscures a lot, but it should not be compiled into release version. I have no better answer to this and no better idea too.